### PR TITLE
Basic LPUART support for STM32Lx

### PR DESCRIPTION
**Pull request by tagunil**
Tested on STM32L073.

### CODE REVIEW

**Comment by erwango** [comment]
@vaussard , I can't add you to the review, but would be nice if you can have a look.

File: drivers/serial/Kconfig.stm32 [hunk 1]
```diff
@@ -7,10 +7,12 @@
 
 menuconfig UART_STM32
 	bool "STM32 MCU serial driver"
-	default n
+	depends on SOC_FAMILY_STM32
 	select SERIAL_HAS_DRIVER
 	select SERIAL_SUPPORT_INTERRUPT
-	depends on SOC_FAMILY_STM32
+	select USE_STM32_LL_USART
```
Comment (vaussard) [inline drivers/serial/Kconfig.stm32]: Selecting USE_STM32_LL_USART does not look related to LPUART. Is that a separate fix?

File: drivers/serial/uart_stm32.c [hunk 2]
```diff
@@ -287,15 +287,28 @@ static int uart_stm32_init(struct device *dev)
...
+#ifdef USART_PRESC_PRESCALER
+				     LL_USART_PRESCALER_DIV1,
 #endif
 #ifdef USART_CR1_OVER8
-			     LL_USART_OVERSAMPLING_16,
+				     LL_USART_OVERSAMPLING_16,
 #endif
-			     config->baud_rate);
+				     config->baud_rate);
+	}
```
Comment (vaussard) [inline drivers/serial/uart_stm32.c]: This part starts to be really hard to read due to the excessive ifdef'ery. I had to read twice to understand why the `if (1)` in the middle. What about completely decoupling STM32Lxx from the rest as both parts have few in common?
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || defined(CONFIG_SOC_SERIES_STM32L4X)
    if (IS_LPUART_INSTANCE(UartInstance)) {
        LL_LPUART_SetBaudRate(...);
    }
#else /* CONFIG_SOC_SERIES_STM32Lxx */
    LL_USART_SetBaudRate(...);
#endif
```

File: dts/arm/st/stm32l4.dtsi [hunk 3]
```diff
@@ -90,6 +90,15 @@
 			label = "UART_3";
 		};
 
+		lpuart1: serial@40008000 {
+			compatible = "st,stm32-lpuart";
+			reg = <0x40008000 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00004000>;
```
Comment (vaussard) [inline dts/arm/st/stm32l4.dtsi]: It should be `clocks = <&rcc STM32_CLOCK_BUS_APB1_2 0x00000001>;` as bit 0 is used on L4.

File: dts/arm/st/stm32l0.dtsi [hunk 4]
```diff
@@ -80,6 +80,15 @@
 			label = "UART_2";
 		};
 
+		lpuart1: serial@40004800 {
+			compatible = "st,stm32-lpuart";
```
Comment (vaussard) [inline dts/arm/st/stm32l0.dtsi]: The compatible should probably be `"st,stm32-lpuart", "st,stm32-uart"` as we are using the common driver for all UART flavours. I checked the Linux binding for STM32, but it does not have support for LPUART.

**Review by vaussard — COMMENTED** [review]
Hi @erwango, it is done. I did a similar change myself as while ago to support the LPUART, so it looks good apart a few minor comments.

**Comment by vaussard** [comment]
Mmmh why are some of my comments shown as outdated, while the code did not changed? Strange github.

File: drivers/serial/uart_stm32.c [hunk 5]
```diff
@@ -291,7 +291,7 @@ static int uart_stm32_init(struct device *dev)
 #ifdef USART_CR1_OVER8
 			     LL_USART_OVERSAMPLING_16,
 #endif
-			     data->huart.Init.BaudRate);
+			     config->baud_rate);
```
[the remark this answers is not in the document]
Comment (tagunil) [reply drivers/serial/uart_stm32.c]: No, you are right, that should be done in a separate commit.

File: drivers/serial/Kconfig.stm32 (hunk 1 above)
Comment (tagunil) [reply drivers/serial/Kconfig.stm32]: Yes, it's just a fixing things around. I can separate it too.

File: dts/arm/st/stm32l4.dtsi (hunk 3 above)
Comment (tagunil) [reply dts/arm/st/stm32l4.dtsi]: Thanks, I'll look at that!

File: dts/arm/st/stm32l0.dtsi (hunk 4 above)
Comment (tagunil) [reply dts/arm/st/stm32l0.dtsi]: I'll fix that.

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (tagunil) [reply drivers/serial/uart_stm32.c]: Well, then heavily ifdef'ed code for ordinary USARTs will be duplicated like that:
```
#if defined(CONFIG_SOC_SERIES_STM32L0X) || defined(CONFIG_SOC_SERIES_STM32L4X)
    if (IS_LPUART_INSTANCE(UartInstance)) {
        LL_LPUART_SetBaudRate(...);
    } else {
        LL_USART_SetBaudRate(...);
    }
#else /* CONFIG_SOC_SERIES_STM32Lxx */
    LL_USART_SetBaudRate(...);
#endif
```
Is that really better?

**Comment by tagunil** [comment]
@vaussard, thank you for detailed review!

File: dts/arm/st/stm32l4.dtsi (hunk 3 above)
Comment (tagunil) [reply dts/arm/st/stm32l4.dtsi]: Yes, it's just stupid copy-paste mistakes. I'll fix that.

File: drivers/serial/Kconfig.stm32 (hunk 1 above)
Comment (tagunil) [reply drivers/serial/Kconfig.stm32]: Though it should be fixed before adding LPUART.

File: dts/arm/st/stm32l0-pinctrl.dtsi [hunk 6]
```diff
@@ -21,6 +21,12 @@
 					tx = <STM32_PIN_PA2 (STM32_PINMUX_ALT_FUNC_4 | STM32_PUSHPULL_NOPULL)>;
 				};
 			};
+			lpuart1_pins_a: lpuart1@0 {
```
Comment (erwango) [inline dts/arm/st/stm32l0-pinctrl.dtsi]: Unfortunately, pin generation is not there yet, so to be actually tested, updating boards pinmux.c will be required for now.

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (erwango) [reply drivers/serial/uart_stm32.c]: Another option is to add use a new function either in this file, either in new file uart_lx_stm32.c

File: dts/arm/st/stm32l0-pinctrl.dtsi (hunk 6 above)
Comment (tagunil) [reply dts/arm/st/stm32l0-pinctrl.dtsi]: I've done that in my working copy to test. Nucleo-L073RZ is good for such testing, because LPUART1 is muxed with USART2. I don't think however that  some board pinmux should be modified permanently.

File: drivers/serial/uart_stm32.c (hunk 2 above)
Comment (tagunil) [reply drivers/serial/uart_stm32.c]: I've tried to make this thing more readable in today's push. Could you look at that?

Comment (vaussard) [reply drivers/serial/uart_stm32.c]: It is far more readable, thank you.

File: drivers/serial/Kconfig.stm32 (hunk 1 above)
Comment (erwango) [reply drivers/serial/Kconfig.stm32]: I don't think this was actually required for U(S)ART driver.
Stm32 serial driver only used LL headers until now, which do not require this flag.

File: drivers/serial/Kconfig.stm32 [hunk 7]
```diff
@@ -7,10 +7,11 @@
...
```
Comment (erwango) [inline drivers/serial/Kconfig.stm32]: I don't think this was actually required for U(S)ART driver which actually compiled w/o this flag activated.
STM32 serial driver only used LL headers until now, which does not require this flag.
Please re-check and remove

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/serial/Kconfig.stm32 (hunk 7 above)
Comment (tagunil) [reply drivers/serial/Kconfig.stm32]: I see. Headers are sufficient. Will do.

Comment (tagunil) [reply drivers/serial/Kconfig.stm32]: Done.

**Comment by tagunil** [comment]
@erwango, could you look again?

**Review by erwango — APPROVED** [review]

**Review by galak — APPROVED** [review]
